util/udt_util_test.cc

```diff
@@ -0,0 +1,336 @@
+
+    Status MarkRollback(const Slice&) override { return Status::OK(); }
+
+    Status MarkCommit(const Slice&) override { return Status::OK(); }
+
+    Status MarkCommitWithTimestamp(const Slice&, const Slice&) override {
+      return Status::OK();
+    }
+
+    Status MarkNoop(bool) override { return Status::OK(); }
+
+    const std::vector<std::pair<uint32_t, const Slice>>& GetKeys() const {
+      return keys_;
+    }
+
+   private:
+    Status AddKey(uint32_t cf, const Slice& key) {
+      keys_.push_back(std::make_pair(cf, key));
+      return Status::OK();
+    }
+    std::vector<std::pair<uint32_t, const Slice>> keys_;
+  };
+
+  void CreateKey(std::string* key_buf, size_t ts_sz) {
+    if (ts_sz > 0) {
+      AppendKeyWithMinTimestamp(key_buf, kTestKeyWithoutTs, ts_sz);
+    } else {
+      key_buf->assign(kTestKeyWithoutTs);
+    }
+  }
+
+  void CreateWriteBatch(
+      const std::unordered_map<uint32_t, size_t>& ts_sz_for_batch,
+      std::unique_ptr<WriteBatch>& batch) {
+    for (const auto& [cf_id, ts_sz] : ts_sz_for_batch) {
+      std::string key;
+      CreateKey(&key, ts_sz);
+      ASSERT_OK(
+          WriteBatchInternal::Put(batch.get(), cf_id, key, kValuePlaceHolder));
+      ASSERT_OK(WriteBatchInternal::Delete(batch.get(), cf_id, key));
+      ASSERT_OK(WriteBatchInternal::SingleDelete(batch.get(), cf_id, key));
+      ASSERT_OK(WriteBatchInternal::DeleteRange(batch.get(), cf_id, key, key));
+      ASSERT_OK(WriteBatchInternal::Merge(batch.get(), cf_id, key,
+                                          kValuePlaceHolder));
+      ASSERT_OK(WriteBatchInternal::PutBlobIndex(batch.get(), cf_id, key,
+                                                 kValuePlaceHolder));
+    }
+  }
+
+  void CheckSequenceEqual(const WriteBatch& orig_batch,
+                          const WriteBatch& new_batch) {
+    ASSERT_EQ(WriteBatchInternal::Sequence(&orig_batch),
+              WriteBatchInternal::Sequence(&new_batch));
+  }
+  void CheckCountEqual(const WriteBatch& orig_batch,
+                       const WriteBatch& new_batch) {
+    ASSERT_EQ(WriteBatchInternal::Count(&orig_batch),
+              WriteBatchInternal::Count(&new_batch));
+  }
+
+  void VerifyKeys(
+      const std::vector<std::pair<uint32_t, const Slice>>& keys_with_ts,
+      const std::vector<std::pair<uint32_t, const Slice>>& keys_without_ts,
+      size_t ts_sz, std::optional<uint32_t> dropped_cf) {
+    ASSERT_EQ(keys_with_ts.size(), keys_without_ts.size());
+    const std::string kTsMin(ts_sz, static_cast<unsigned char>(0));
+    for (size_t i = 0; i < keys_with_ts.size(); i++) {
+      // TimestampRecoveryHandler ignores dropped column family and copy it over
+      // as is. Check the keys stay the same.
+      if (dropped_cf.has_value() &&
+          keys_with_ts[i].first == dropped_cf.value()) {
+        ASSERT_EQ(keys_with_ts[i].first, keys_without_ts[i].first);
+        ASSERT_EQ(keys_with_ts[i].second.compare(keys_without_ts[i].second), 0);
```

Comment:
`Slice` has `operator==` implemented to compare the contents, so you could just do `ASSERT_EQ(keys_with_ts[i].second, keys_without_ts[i].second)`

Reply:
Thank you for mentioning this, that operator one looks more native!